### PR TITLE
seccomp: whitelist io-uring related system calls

### LINKED ISSUE #39396: io-uring support
I'm currently using containers for my workzone, and I'm just getting my hands on the new io-uring thingy in kernel 5.1 [here](http://git.kernel.dk/cgit/liburing/)

And then I noticed that docker currently blocks the io-uring related requests :

```
/code/liburing/examples on  master! ⌚ 10:14:17
$ ./io_uring-test io_uring-test.c
queue_init: Operation not permitted
```

While this test program runs without a problem on the host machine (I'm using a fedora30 host with docker-machine --drive generic)

So I thought this would be the place to suggest this feature

### PR DESCRIPTION
**Pull request by omegacoleman**
closes #39396

**- What I did**

I added the io-uring related system call introduced in kernel 5.1 to the seccomp whitelist. 

**- How I did it**

This is merely a follow-up to the [upstream](https://github.com/seccomp/libseccomp/commit/5fc22428507ecea00ee9e2215d972777da9a99b6)

**- How to verify it**

Build docker with the master branch libseccomp, and will be run programs using io-uring without trouble.

With older kernels or older versions of libseccomp, the configure will be omitted.

**- Description for the changelog**

seccomp: whitelist io-uring related system calls

**- A picture of a cute animal (not mandatory but encouraged)**

Would be a waste for such a simple change..

### CODE REVIEW

**Comment by thaJeztah** [comment]
Looks like CI is failing, because the json version was not updated as part of this PR;

```
17:32:36 The result of go generate ./profiles/seccomp/ differs
17:32:36 
17:32:36  M profiles/seccomp/default.json
17:32:36 
17:32:36 Please re-run go generate ./profiles/seccomp/
```

**Comment by thaJeztah** [comment]
ping @justincormack @crosbymichael PTAL

**Comment by omegacoleman** [comment]
> 
> 
> Looks like CI is failing, because the json version was not updated as part of this PR;
> 
> ```
> 17:32:36 The result of go generate ./profiles/seccomp/ differs
> 17:32:36 
> 17:32:36  M profiles/seccomp/default.json
> 17:32:36 
> 17:32:36 Please re-run go generate ./profiles/seccomp/
> ```

I made a mistake thinking the json file is manually written, and it should be generated. I have fixed that now but some of the ci seems to be failing because of removing some non-existing folder (C:\go), what does that mean?

**Comment by thaJeztah** [comment]
Thanks! The RS5 failure can be ignored (see https://github.com/moby/moby/pull/39403)

**Comment by omegacoleman** [comment]
> 
> 
> Thanks! The RS5 failure can be ignored (see #39403)

The only failure left is TestAPISwarmServicesMultipleAgents in janky, which does not seem to be relevant. I read the code but it is not about seccomp.

**Comment by thaJeztah** [comment]
yes, looks like that one is sometimes flaky; https://github.com/moby/moby/issues/23626

I'll restart CI

**Comment by omegacoleman** [comment]
> 
> 
> yes, looks like that one is sometimes flaky; #23626
> 
> I'll restart CI

This time it's only a few SKIPs & XFAILs, why is it marked as failing again?

**Comment by thaJeztah** [comment]
Looks to be an issue with the docker-py tests;

```
15:06:25 _______________________ ExecTest.test_exec_command_demux _______________________
15:06:25 /docker-py/tests/integration/api_exec_test.py:182: in test_exec_command_demux
15:06:25     assert next(exec_log) == (b'hello out\r\n', None)
15:06:25 E   AssertionError: assert ('hello out\r...rr\r\n', None) == ('hello out\r\n', None)
15:06:25 E     At index 0 diff: 'hello out\r\nhello err\r\n' != 'hello out\r\n'
15:06:25 E     Use -v to get the full diff
15:06:25 - generated xml file: /go/src/github.com/docker/docker/bundles/test-docker-py/results.xml -
```

I'll kick that build

File: profiles/seccomp/seccomp_default.go [hunk 1]
```diff
@@ -160,6 +160,9 @@ func DefaultProfile() *types.Seccomp {
 				"ioprio_set",
 				"io_setup",
 				"io_submit",
+				"io_uring_setup",
+				"io_uring_enter",
+				"io_uring_register",
```
Comment (justincormack) [inline profiles/seccomp/seccomp_default.go]: Can you sort these alphabetically like the rest please

**Comment by justincormack** [comment]
Are there any other new syscalls since the last update?

**Comment by thaJeztah** [comment]
ping @omegacoleman could you answer @justincormack's question (and sort the list https://github.com/moby/moby/pull/39415#discussion_r298995826)?

**Comment by omegacoleman** [comment]
> Are there any other new syscalls since the last update?

The precise answer is "I don't know" since I'm just a network dev not kernel dev, but even though there are, the new system call has to be added to the upstream libseccomp's syscall translation (name -> number) before it could be added to the config here, and these calls were added.

Sorry for the delay as it is well, trival & as the upstream has added the calls I was able to just override the config myself..

File: profiles/seccomp/seccomp_default.go (hunk 1 above)
Comment (omegacoleman) [reply profiles/seccomp/seccomp_default.go]: done that

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks for updating; could you try rebasing the PR to get rid of the merge commit?

**Comment by omegacoleman** [comment]
> Thanks for updating; could you try rebasing the PR to get rid of the merge commit?

done

**Review by justincormack — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by thaJeztah** [comment]
windows failure is unrelated; let me get this one in

**Comment by cyphar** [comment]
This seems like it's a bad idea -- io_uring allows programs to execute certain syscalls without being limited by seccomp. Currently all of the syscalls you can access through io_uring are permitted in the default profile, but I'm a little worried that this wasn't brought up at the time.

**Comment by justincormack** [comment]
@cyphar it looks like `io_uring` will get its own restriction mechanisms see eg https://lwn.net/SubscriberLink/826053/c58110f9c3493661/. I really don't think we can restrict access to a syscall that lots of people want to use based on the fact it might in future add support for unsafe actions.

I am reviewing the whole seccomp policy based on what has been useful in the past.

**Comment by cyphar** [comment]
Funnily enough, I found this PR after commenting on that LWN article. The problem with that restriction mechanism is that it doesn't really fulfil the seccomp usecase (as long as you can create a new `io_uring` ring, you can opt to not enable restrictions for it). But Kees has said he's interested in figuring out how to get seccomp to limit `io_uring`, so I am hopeful this will happen.

**Comment by omegacoleman** [comment]
> Funnily enough, I found this PR after commenting on that LWN article. The problem with that restriction mechanism is that it doesn't really fulfil the seccomp usecase (as long as you can create a new `io_uring` ring, you can opt to not enable restrictions for it). But Kees has said he's interested in figuring out how to get seccomp to limit `io_uring`, so I am hopeful this will happen.

From what I got from the news they're working on sth like using eBPF to filter the SQEs upon ktail update. Will try to make that configurable after kernel & libseccomp get that feature.

**Comment by AkihiroSuda** [comment]
Shall we revert this until we find the solution?

**Comment by cyphar** [comment]
I don't think it's necessary to revert. The default profile allows everything that you can do with `io_uring` -- and custom profiles should be modified accordingly. I was more just surprised that this wasn't mentioned at the time.

**Comment by thaJeztah** [comment]
for future visitors; these were removed again in;

- https://github.com/moby/moby/pull/46762
- and in containerd: https://github.com/containerd/containerd/pull/9320

**Pull request merged**